From 1ef45859ac39bc01e5b5903ace0fd3f659dbee7b Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Thu, 11 Jun 2026 17:54:03 +0800
Subject: [PATCH 45/48] loonggpu: Suspend and resume clients with client
 helpers

Replace calls to radeon_fbdev_set_suspend() with calls to the client
functions drm_client_dev_suspend() and drm_client_dev_resume(). Any
registered in-kernel client will now receive suspend and resume events.

Following the reference implementation changes.

Link: https://git.kernel.org/torvalds/c/88c79de8575c
Link: https://git.kernel.org/torvalds/c/612ec7c69d04
Link: https://git.kernel.org/torvalds/c/fff8e0504499
Link: https://git.kernel.org/torvalds/c/7910d69376cd
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 include/loonggpu.h         |  4 ++--
 loonggpu/loonggpu_device.c | 19 ++++++++++++-------
 2 files changed, 14 insertions(+), 9 deletions(-)

diff --git a/include/loonggpu.h b/include/loonggpu.h
index 129e25a..eb5d208 100644
--- a/include/loonggpu.h
+++ b/include/loonggpu.h
@@ -1725,8 +1725,8 @@ int loonggpu_driver_open_kms(struct drm_device *dev, struct drm_file *file_priv)
 void loonggpu_driver_postclose_kms(struct drm_device *dev,
 				 struct drm_file *file_priv);
 int loonggpu_device_ip_suspend(struct loonggpu_device *adev);
-int loonggpu_device_suspend(struct drm_device *dev, bool suspend, bool fbcon);
-int loonggpu_device_resume(struct drm_device *dev, bool resume, bool fbcon);
+int loonggpu_device_suspend(struct drm_device *dev, bool suspend, bool notify_clients);
+int loonggpu_device_resume(struct drm_device *dev, bool resume, bool notify_clients);
 u32 loonggpu_get_vblank_counter_kms(struct drm_device *dev, unsigned int pipe);
 long loonggpu_kms_compat_ioctl(struct file *filp, unsigned int cmd,
 			     unsigned long arg);
diff --git a/loonggpu/loonggpu_device.c b/loonggpu/loonggpu_device.c
index e8bbc8d..743a0f6 100644
--- a/loonggpu/loonggpu_device.c
+++ b/loonggpu/loonggpu_device.c
@@ -1,10 +1,10 @@
 #include <linux/power_supply.h>
 #include <linux/kthread.h>
-#include <linux/console.h>
 #include <linux/slab.h>
 #include "loonggpu.h"
 #include <drm/drm_crtc_helper.h>
 #include <drm/drm_atomic_helper.h>
+#include <drm/drm_client_event.h>
 #include "loonggpu_drm.h"
 #include <linux/vgaarb.h>
 #include <linux/vga_switcheroo.h>
@@ -1714,6 +1714,11 @@ static int loonggpu_zip_gem_bo_evict(int id, void *ptr, void *data)
 	return 0;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 19, 0)
+#define drm_client_dev_suspend(x) (drm_client_dev_suspend)((x), false)
+#define drm_client_dev_resume(x) (drm_client_dev_resume)((x), false)
+#endif
+
 /*
  * Suspend & resume.
  */
@@ -1728,7 +1733,7 @@ static int loonggpu_zip_gem_bo_evict(int id, void *ptr, void *data)
  * Returns 0 for success or an error on failure.
  * Called at driver suspend.
  */
-int loonggpu_device_suspend(struct drm_device *dev, bool suspend, bool fbcon)
+int loonggpu_device_suspend(struct drm_device *dev, bool suspend, bool notify_clients)
 {
 	struct loonggpu_device *adev;
 	int r;
@@ -1745,8 +1750,8 @@ int loonggpu_device_suspend(struct drm_device *dev, bool suspend, bool fbcon)
 
 	drm_kms_helper_poll_disable(dev);
 
-	if (fbcon)
-		loonggpu_fbdev_set_suspend(adev, 1);
+	if (notify_clients)
+		drm_client_dev_suspend(dev);
 
 	loonggpu_lgkcd_suspend(adev, false);
 
@@ -1802,7 +1807,7 @@ int loonggpu_device_suspend(struct drm_device *dev, bool suspend, bool fbcon)
  * Returns 0 for success or an error on failure.
  * Called at driver resume.
  */
-int loonggpu_device_resume(struct drm_device *dev, bool resume, bool fbcon)
+int loonggpu_device_resume(struct drm_device *dev, bool resume, bool notify_clients)
 {
 	struct loonggpu_device *adev = dev->dev_private;
 	struct loonggpu_dc *dc = adev->dc;
@@ -1883,8 +1888,8 @@ int loonggpu_device_resume(struct drm_device *dev, bool resume, bool fbcon)
 	flush_delayed_work(&adev->late_init_work);
 
 	/* blat the mode back in */
-	if (fbcon)
-		loonggpu_fbdev_set_suspend(adev, 0);
+	if (notify_clients)
+		drm_client_dev_resume(dev);
 
 	drm_kms_helper_poll_enable(dev);
 
-- 
2.52.0

